server/src/controllers/course.ts

```diff
@@ -90,19 +101,105 @@ function parseSisuInstance(instance: SisuInstance): InstanceData {
   };
 }
 
+// Middleware function for handling errors in JSON parsing
+export function handleInvalidRequestJson(err: JsonError, req: Request, res: Response, next: NextFunction): void {
+  if (err instanceof SyntaxError && err.status === 400 && err.body !== undefined) {
+    res.status(400).send({
+      success: false,
+      errors: [ `SyntaxError: ${err.message}: ${err.body}` ]
+    });
+  } else
+    next();
+}
 
 export async function addCourse(req: Request, res: Response): Promise<void> {
+  const requestSchema: yup.AnyObjectSchema = yup.object().shape({
+    courseCode: yup.string().required(),
+    translations: yup.array().of(yup.object().shape({
+      lang: yup.mixed<Language>().oneOf(Object.values(Language)).required(),
+      department: yup.string().required(),
+      name: yup.string().required()
+    })).required()
+  });
+
+  const t: Transaction = await sequelize.transaction();
+
   try {
-    // TODO: add the course to the database
-    res.send({
-      success: true
+    await requestSchema.validate(req.body, { abortEarly: false });
+
+    const course: Course = await Course.create({
+      courseCode: req.body.courseCode
     });
-  } catch (error) {
-    res.status(401);
-    res.send({
-      success: false,
-      error: error,
+
+    const courseData: CourseData = {
+      id: course.id,
+      courseCode: course.courseCode,
+      department: {
+        en: '',
+        fi: '',
+        sv: ''
+      },
+      name: {
+        en: '',
+        fi: '',
+        sv: ''
+      },
+      evaluationInformation: {
+        en: '',
+        fi: '',
+        sv: ''
+      }
+    };
+
+    for (const translation of req.body.translations) {
```

Comment:
Instead of using db calls in the for loop to create the translations, could we utilize the sequelize [bulkCreate](https://sequelize.org/api/v6/class/src/model.js~model#static-method-bulkCreate)?

Could be maybe something like:
```JavaScript
await CourseTranslation.bulkCreate([
  { /*FI translations*/ },
  { /*EN translations*/ },
  { /*SV translations*/ }
])
```

I think the performance of this route is not so critical because new courses are not added that frequently. In this case looping few database calls would not pose any performance issues (+ its run in a transaction anyway). Its just good to note that sequelize has this kind of method also.

I'll let you decide what to do.

Reply:
I could probably do the db operations in bulkCreate and then loop through the array to do the rest of the operations.